crates/shared/src/price_estimation/trade_finder.rs

```diff
@@ -119,126 +148,211 @@ impl Inner {
+        OrderKind::Sell => vec![trade.out_amount, query.in_amount],
+        OrderKind::Buy => vec![query.in_amount, trade.out_amount],
+    };
+
+    // Configure the most disadvantageous trade possible. Should the trader not
+    // receive the amount promised by the [`Trade`] the simulation will still work
+    // and we can compute the actual [`Trade::out_amount`] afterwards.
+    let (sell_amount, buy_amount) = match query.kind {
+        OrderKind::Sell => (query.in_amount, 0.into()),
+        OrderKind::Buy => (U256::MAX, query.in_amount),
+    };
+    let fake_order = OrderData {
+        sell_token: query.sell_token,
+        sell_amount,
+        buy_token: query.buy_token,
+        buy_amount,
+        receiver: Some(query.receiver),
+        valid_to: u32::MAX,
+        app_data: Default::default(),
+        fee_amount: 0.into(),
+        kind: query.kind,
+        partially_fillable: false,
+        sell_token_balance: query.sell_token_source,
+        buy_token_balance: query.buy_token_destination,
+    };
+
+    let fake_signature = Signature::default_with(SigningScheme::Eip1271);
+    let encoded_trade = encode_trade(
+        &fake_order,
+        &fake_signature,
+        query.from,
+        0,
+        1,
+        &query.in_amount,
+    );
+
+    EncodedSettlement {
+        tokens,
+        clearing_prices,
+        trades: vec![encoded_trade],
+        interactions: [
+            encode_interactions(&query.pre_interactions),
+            trade_interactions,
+            encode_interactions(&query.post_interactions),
+        ],
+    }
+}
+
+/// Adds the interactions that are only needed to query important balances
+/// throughout the simulation.
+/// These balances will get used to compute an accurate price for the trade.
+fn add_balance_queries(
+    mut settlement: EncodedSettlement,
+    query: &QuoteQuery,
+    settlement_contract: H160,
+    solver: &Solver,
+) -> EncodedSettlement {
+    let (token, owner) = match query.kind {
+        // track how much `buy_token` the `receiver` actually got
+        OrderKind::Sell => (query.buy_token, query.receiver),
+        // track how much `sell_token` the settlement contract actually spent
+        OrderKind::Buy => (query.sell_token, settlement_contract),
+    };
+    let query_balance = solver.methods().store_balance(token, owner);
+    let query_balance = Bytes(query_balance.tx.data.unwrap().0);
+    let interaction = (solver.address(), 0.into(), query_balance);
+    settlement.interactions[1].insert(0, interaction.clone());
+    match query.kind {
+        // query `receiver` balance right after paying out funds (first post-interaction)
+        OrderKind::Sell => settlement.interactions[2].insert(0, interaction),
+        // query `settlement` balance right before paying out funds (last regular interaction)
+        OrderKind::Buy => settlement.interactions[1].push(interaction),
+    }
+    settlement
+}
+
 impl TradeVerifier {
     const DEFAULT_GAS: u64 = 8_000_000;
```

Comment:
Lets use the block gas limit. This caused issues before:
```suggestion
    const DEFAULT_GAS: u64 = 30_000_000;
```

Reply:
Actually nevermind. This isn't part of this PR and using the block gas limit has its own problems.